File: content/problems/0622-design-circular-queue/solutions/array-ring-buffer/Solution.java
class MyCircularQueue {
    private final int[] values;
    private int head,size;
    public MyCircularQueue(int k){
        values=new int[k];
    }
    public boolean enQueue(int value){
        if(isFull())return false;
        values[(head+size)%values.length]=value;
        size++;
        return true;
    }
    public boolean deQueue(){
        if(isEmpty())return false;
        head=(head+1)%values.length;
        size--;
        return true;
    }
    public int Front(){
        return isEmpty()?-1:values[head];
    }
    public int Rear(){
        return isEmpty()?-1:values[(head+size-1)%values.length];
    }
    public boolean isEmpty(){
        return size==0;
    }
    public boolean isFull(){
        return size==values.length;
    }
}

File: content/problems/0622-design-circular-queue/solutions/array-ring-buffer/Solution.kt
class MyCircularQueue(k:Int){
    private val values=IntArray(k);
    private var head=0;
    private var size=0;
    fun enQueue(value:Int):Boolean{
        if(isFull())return false;
        values[(head+size)%values.size]=value;
        size++;
        return true
    };
    fun deQueue():Boolean{
        if(isEmpty())return false;
        head=(head+1)%values.size;
        size--;
        return true
    };
    fun Front():Int=if(isEmpty())-1 else values[head];
    fun Rear():Int=if(isEmpty())-1 else values[(head+size-1)%values.size];
    fun isEmpty():Boolean=size==0;
    fun isFull():Boolean=size==values.size
}
